compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -843,13 +843,25 @@ impl<'hir> LoweringContext<'_, 'hir> {
         // mut __awaitee => loop { ... }
         let awaitee_arm = self.arm(awaitee_pat, loop_expr);
 
-        // `match ::std::future::IntoFuture::into_future(<expr>) { ... }`
-        let into_future_expr = self.expr_call_lang_item_fn(
-            span,
-            hir::LangItem::IntoFutureIntoFuture,
-            arena_vec![self; expr],
-            Some(expr_hir_id),
+        // `match <expr>.into_future() { ... }`
+        let into_future_path = self.arena.alloc(hir::PathSegment {
+            ident: Ident::new(sym::into_future, self.lower_span(span)),
+            hir_id: expr_hir_id,
+            res: hir::PathSegmentRes::LangItem(
+                hir::def::DefKind::AssocFn,
+                hir::LangItem::IntoFutureIntoFuture,
+            ),
```

Comment:
Why do we need this? Why can't we just use `Res::Def` by itself?

```suggestion
            res: Res::Def(DefKind::AssocFn, tcx.require_lang_item(hir::LangItem::IntoFutureIntoFuture, Some(span))),
```

Reply:
This would only break if, e.g., we had an `.await` call inside of core itself, which I guess may happen. But this really seems complicated for such a special use case.

Reply:
Using `Res::def` directly was my initial approach, but the compiler would crash when trying to get the lang items. I didn't investigate in detail, but I think lang items just aren't available at that point,

Reply:
Sigh, there are calls to await in core.